File: homework1/src/warmup.js
// HOMEWORK 1, SEBASTIAN GRASSO

function change(cents) {
  if (cents < 0) {
    throw new RangeError('RangeError: amount cannot be negative');
  }

  let remaining = cents;

  const quarters = Math.floor(cents / 25);
  remaining %= 25;
  const dimes = Math.floor(remaining / 10);
  remaining %= 10;
  const nickels = Math.floor(remaining / 5);
  remaining %= 5;
  const pennies = remaining;

  return [quarters, dimes, nickels, pennies];
}

function stripQuotes(string) {
  return string.replace(/['"]+/g, '');
}

function scramble(s) {
  const seperate = s.split('');
  const letters = seperate.length - 1;
  for (let i = letters; i > 0; i -= 1) {
    const temp = seperate[i];
    const randSpot = Math.floor(Math.random() * (i + 1));
    seperate[i] = seperate[randSpot];
    seperate[randSpot] = temp;
  }
  return seperate.join('');
}

function powers(base, limit, callback) {
  let exp = 0;
  while (base ** exp <= limit) {
    callback(base ** exp);
    exp += 1;
  }
}

// credit DR TOALS INTROUCTION TO JAVASCRIPT (https://cs.lmu.edu/~ray/notes/javascript/)
function* powersGenerator(base, limit) {
  let value = 1;
  while (value <= limit) {
    yield value;
    value *= base;
  }
}

function say(word) {
  return nextWord(word);
//   if (word === undefined){
//     return ''
//   }
//   else {
//     const nextWord = undefined;
//       if (!nextWord) {
//         nextWord = word;
//       }
//       else{
//     }
//
// function
//   }
}

function interleave(arr, ...values) {
  const weave = [];

  for (let a = 0; a < Math.max(arr.length, values.length); a += 1) {
    if (a < arr.length) { weave.push(arr[a]); }
    if (a < values.length) { weave.push(values[a]); }
  }
  return weave;
}
// credit DR TOALS INTROUCTION TO JAVASCRIPT (https://cs.lmu.edu/~ray/notes/javascript/)
function cylinder(spec) {
  let { radius = 1, height = 1 } = spec;

  const surfaceArea = () => (2 * Math.PI * radius * radius) + (2 * Math.PI * radius * height);
  const volume = () => Math.PI * radius * radius * height;
  const stretch = (factor) => { height *= factor; };
  const widen = (factor) => { radius *= factor; };

  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    volume,
    surfaceArea,
    get radius() { return radius; },
    get height() { return height; },
    toString,
    stretch,
    widen,
  });
}

function makeCryptoFunctions(key, algo) {
  const crypto = require('crypto');

  function encrypt(str) {
    const alpha = crypto.createCipher(algo, key);
    const encoded = alpha.update(str, 'utf8', 'hex') + alpha.final('hex');
    return encoded;
  }

  function decrypt(str) {
    const alpha = crypto.createDecipher(algo, key);
    const decoded = alpha.update(str, 'hex', 'utf8') + alpha.final('utf8');
    return decoded;
  }

  return [encrypt, decrypt];
}

function randomName({ region, gender }) {
  const prom = require('request-promise');
  const query = {
    uri: 'https://uinames.com/api/',
    qs: {
      gender,
      region,
    },
    json: true,
  };
  return prom(query).then((body) => `${body.name}, ${body.surname}`);
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  say,
  powersGenerator,
  interleave,
  cylinder,
  makeCryptoFunctions,
  randomName,
};

File: homework2/warmup.py
#HomeWork 2, Sebastian Grasso

import math
from cryptography.fernet import Fernet
import re
import requests
import random

#Problem 1
def change(number):
    if number < 0:
        raise ValueError('amount cannot be negative')
        
    denominations = {'Quarter': 25, 'Dime': 10, 'Nickel': 5, 'Penny': 1}

    quarters = math.floor(number/denominations['Quarter'])
    number %= denominations['Quarter']
    dimes = math.floor(number / denominations['Dime'])
    number %= denominations['Dime']
    nickels = math.floor(number / denominations['Nickel'])
    number %= denominations['Nickel']
    pennies = math.floor(number / denominations['Penny'])
    number %= denominations['Penny']

    return quarters, dimes, nickels, pennies

#Problem 2
def strip_quotes(string):
    return string.replace('\'', '').replace('\"', "")

#Problem 3
def scramble(string):
    return ''.join(random.sample(string, len(string)))

#Problem 4
def powers(base, limit):
    successive = 1
    while successive <= limit:
        yield successive
        successive *= base

#Problem 5
def triples(value):
    triplelist = list()
    for a in range(1, value+1):
        for b in range(1, value+1):
            for c in range(1, value+1):
                if (a ** 2) + (b ** 2) == (c ** 2) and (a < b < c):
                    tripleList.append(a,b,c)
    return triplelist

def say(firstWord = ''):
    if firstWord == '':
        return firstWord

    def sayAgain(nextWord = ''):
        if nextWord == '':
            return firstWord
        return say(firstWord + ' ' + nextWord)

    return sayAgain

#7
def interleave(array, *argv):
    result = []
    arrayLength = len(array)
    valuesLength = 0
    for arg in argv:
        valuesLength+=1
    max_len = max(arrayLength, valuesLength)
    for i in range(max_len):
        if (i < arrayLength):
            result.append(array[i])
        if (i < valuesLength):
            result.append(argv[i])
    return result

#8
class Cylinder:
    def __init__(self, radius = 1, height = 1):
        self.radius = radius
        self.height = height
    @property
    def volume(self):
        return math.pi * self.radius * self.radius * self.height
    @property
    def surface_area(self):
        return 2 * math.pi * self.radius * self.height + (2 * math.pi * self.radius * self.radius)
    def stretch(self, stretch_factor):
        self.height *= stretch_factor
    def widen(self, widen_factor):
        self.radius *= widen_factor


#9
def make_crypto_functions(key):
    k = Fernet(key)
    def encrypt(s):
        return k.encrypt(s)
    def decrypt(token):
        return k.decrypt(token)
    return (encrypt, decrypt)

#10
def random_name(**k):
    url = 'http://uinames.com/api/'
    info = requests.get(url=url, params=k)
    person = json.loads(info.content)
    if 'error' in person:
        raise ValueError('{"error": "Invalid gender"}')
    result = person.get('surname') + ', ' + person.get('name')
    return result
